update add and remove method to care of duplicate elements in graph

MridulS opened the pull request:
This is great! Just started playing around with the library.

While creating graphs there is no check to see if nodes/edges are already created or not, which creates multiple copies of the same node in the Graph object. I have added some checks to take care of this.

I still need to implement the `clear_graph` method which removes everything, so we can update the graph to be a smaller graph than before.

Successive use of
```
cytoscapeobj.graph.add_graph_from_networkx(nx.complete_graph(5))
cytoscapeobj.graph.add_graph_from_networkx(nx.complete_graph(3))
```

will not update the widget to show a complete graph of 3 as nodes and edges from the previous bigger graph needs to be removed first.

marimeireles commented on ipycytoscape/cytoscape.py:
```diff
@@ -155,7 +159,22 @@ def add_edge(self, edge):
+        except KeyError:
+            self.edges.append(edge)
+            if source not in self._adj:
+                node_instance_source = Node()
```
What is the purpose of `node_instance_source` and `node_instance_target`?

MridulS replied:
Whoops, thanks for the catch, should have just been `node_instance`

MridulS commented:
Thanks for the feedback!

> I was thinking that maybe we could change the API and rename add_node method that doesn't really mean anything, to append_node.
> And then we could create a new method for the stuff you wrote called update_node. I think it will be more meaningful for python developers. What do you think? :)

I think sticking with `add_node` is better as `add_` seems to be a "standard" across graph manipulation libraries, networkx, CuGraph, igraph.

From a users POV I think successive `add_graph_from_*` shouldn't mean appending the node/edges to the current graph but rather creating new graphs. User should use something like `add/remove_node/edge` to further manipulate the graph.

We can definitely move the check code to something called `update_node` and every call will check if the nodes/edges are already present.
[PS: My views are probably biased due to working too much with the networkx API, which I really like :) ]

MridulS commented on ipycytoscape/cytoscape.py:
```diff
@@ -191,15 +216,16 @@ def add_graph_from_networkx(self, g):
             receives a generic NetworkX graph. more info in
             https://networkx.github.io/documentation/
         """
+        self.clear_graph()
```
yeah, true, maybe we can a have a method argument like `append_graph=False` which clears up the graph and `append_graph=True` to adds a new graph. 

Also then we should look into keeping a track of unique ids of nodes and edges. For example successive `add_graph_from_*` will add multiple nodes with same id.

marimeireles replied:
Hum, I'm not sure if I follow you here.

If a user wants to clear a graph they can just call `clear()` in the graph and if they want to append new things to a graph they can just use the method `append_node` that's going to be created, no?

MridulS replied:
If the user calls successive `add_graph_from_*` it **should** overwrite the previous object. Yes we can assume that the user has called `clear()` method but I guess its better to have to strict behaviour in terms of what happens when successive calls of `add_graph_from_*` are made.

For the following bit of code the result should be a graph with 3 nodes, not a graph with 8 nodes.
```
>>> obj.add_graph_from_networkx(nx.complete_graph(5))
>>> obj.add_graph_from_networkx(nx.complete_graph(3))
```

The following bit of code, will append the graph, i.e in total there should be 8 nodes.
```
>>> obj.add_graph_from_networkx(nx.complete_graph(5))
>>> obj.add_graph_from_networkx(nx.complete_graph(3), append_graph=True)
```

Let me know if this sounds good :)

marimeireles replied:
>The following bit of code, will append the graph, i.e in total there should be 8 nodes.

Aren't your changes preventing exactly this kind of thing from happening? Having  duplicated node ids. 
But I see your point, if they were different nodes, then calling `add_graph_from_*` more than once in the same graph would append things to a graph.

From my pov if a person wants a new graph they will just create a new `ipycytoscape` object and do a new thing with it. If they're calling `add` in a graph that was already there I think their point is primarily to append things to it.

marimeireles commented:
> I think sticking with add_node is better as add_ seems to be a "standard" across graph manipulation libraries, networkx, CuGraph, igraph.

Okay, that's a very good point. I did a bit of research and seems good. I think it makes more sense to follow the flux of what graphs libs are doing than what python does.
So let's keep the way it is. It's necessary to change the docs to reflect your changes, (the comments in the code). I can do it in a next PR, or if you're interested, please feel free to do it too :)
I'll also add an `append_node()` method, just so we can have it documented and people know that they can use it.

> [PS: My views are probably biased due to working too much with the networkx API, which I really like :) ]

I'm okay with that! hehe

MridulS commented:
addresses issues with https://github.com/QuantStack/ipycytoscape/issues/103

An example of what's improved with this
![example](https://user-images.githubusercontent.com/5363860/85457953-3a045e80-b5be-11ea-818e-7b99fa0a535d.gif)

MridulS commented:
@marimeireles can you review this again? Thanks!

MridulS commented:
overview of changes
- Checks while removing and adding nodes/edges to the graph
- Use add_node/add_edge while creating graphs instead of manipulating Nodes/Edges lists directly
- Fixed the logic in `_by_id` operations.
- add_edge(node1, node2) will work even if they don't exist in the graph, nodes are automatically added
- everything is backed up in `_adj` to maintain a single source of truth. We should use this to sync changes all across.
- better error messages in case elements are not found

ianhi commented:
I'm definitely a fan of being able to do small updates to the graph without having to first create a networkx graph or something similar. 

I haven't looked at this in any detail yet but I like the
> to maintain a single source of truth.

To that point I think that if you're going to create that for python then it would be great to simultaneously build the same on the typescript side. By which I mean that source of truth should be synced to the `model` and then used to generate `views`. I put this thought in more detail in https://github.com/QuantStack/ipycytoscape/issues/96 and I think this PR may be a good place to implement that (conditional on @marimeireles thinking that would be a good change).

Even if no restructuring of the frontend happens it would be good if these `add` and `remove` methods also caused the `cytoscapejs` `add` and `remove` functions to be called (https://js.cytoscape.org/#cy.add). Currently whenever you change one of these methods the entire cytoscape object is scrapped and a new one is created (`init_render` method of the view). Using the cytoscape methods is more parsimonious and would prevent this recreation. The clear benefit to the user of this is that the layout will be preserved (in your gif you can see that the graph briefly disappears and then moves around), though it would like also increase performance especially for large graphs.

marimeireles commented on ipycytoscape/cytoscape.py:
```diff
@@ -137,21 +166,78 @@ def remove_node_by_id(self, node_id):
+                edge_list.append(edge)
+                if source not in self._adj:
+                    node_instance = Node()
+                    # setting the id, according to current spec should be only int/str
```
The thing with setting the `id` to be only either `str/int` is that is not fully compatible with Nx.
I don't see a problem here, the cytoscape docs specify that their `id` is only string anyways.
I just feel like we should document this kind of stuff for the end user. I'll add this to the doc issue, so we don't forget about it.

marimeireles commented:
This PR looks great to me.
I'll rebase it @MridulS if you don't mind so we can merge it fast and I can include it in the next release.
It's so much better to add and remove nodes and edges now, thanks a lot :)
I think it'd be cool to have an example showing how to use the new `add_nodes` and `add_edges` methods, but I'll open an issue for this, we don't have to do it here.

marimeireles commented:
Oh yeah, and the labels example stopped working.
Will have to push a fix here in this PR if you don't mind.

MridulS commented:
fixed the merge conflict

marimeireles commented:
Thanks again for all of your work here @MridulS. It's really great :)
